pipeline-model-declarative-agent/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/DeclarativeAgentDescriptor.java

```diff
@@ -0,0 +1,190 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.plugins.pipeline.modeldefinition.agent;
+
+import hudson.ExtensionComponent;
+import hudson.ExtensionList;
+import hudson.model.Descriptor;
+import org.jenkinsci.plugins.scriptsecurity.sandbox.whitelists.Whitelisted;
+import org.jenkinsci.plugins.structs.SymbolLookup;
+import org.jenkinsci.plugins.structs.describable.DescribableModel;
+import org.jenkinsci.plugins.structs.describable.UninstantiatedDescribable;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Descriptor for {@link DeclarativeAgent}.
+ *
+ * @author Andrew Bayer
+ */
+public abstract class DeclarativeAgentDescriptor extends Descriptor<DeclarativeAgent> {
+
+    /**
+     * The name for this agent type. Should match the {@code Symbol} on the class.
+     *
+     * @return The name.
+     */
+    public abstract @Nonnull String getName();
+
+    /**
+     * The full package and class name for the {@link DeclarativeAgentScript} class corresponding to this.
+     *
+     * @return The class name.
+     */
+    public abstract @Nonnull String getDeclarativeAgentScriptClass();
+
+    /**
+     * Creates an instance of the corresponding {@link DeclarativeAgent} from the given arguments.
+     *
+     * @param arguments A map of strings/objects to be passed to the constructor.
+     * @return An instantiated {@link DeclarativeAgent}
+     * @throws Exception
+     */
+    public DeclarativeAgent newInstance(Map<String,Object> arguments) throws Exception {
+        return new DescribableModel<>(clazz).instantiate(arguments);
+    }
+
+    public static ExtensionList<DeclarativeAgentDescriptor> all() {
+        return ExtensionList.lookup(DeclarativeAgentDescriptor.class);
+    }
+
+    /**
+     * Get a map of name-to-{@link DescribableModel} of all known/registered descriptors.
+     *
+     * @return A map of name-to-{@link DescribableModel}s
+     */
+    public static Map<String,DescribableModel> getDescribableModels() {
+        Map<String,DescribableModel> models = new HashMap<>();
+
+        for (DeclarativeAgentDescriptor d : getOrderedDescriptors()) {
+            Set<String> symbolValues = SymbolLookup.getSymbolValue(d);
+
+            if (!symbolValues.isEmpty()) {
+                models.put(symbolValues.iterator().next(), new DescribableModel<>(d.clazz));
```

Comment:
You should rather loop over all the symbols. Actually easier than what you are doing.